### PR TITLE
[jit] Support pack_padded_sequence and pad_packed_sequence

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23266 [jit] Support nn.GRU in script
* **#23249 [jit] Support pack_padded_sequence and pad_packed_sequence**
* #23248 [jit] support torch._C._get_tracing_state in script
* #23247 [jit] support torch.as_tensor in script

Differential Revision: [D16466587](https://our.internmc.facebook.com/intern/diff/D16466587)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -7736,6 +7736,27 @@ def forward(self, x, seq_lens):
...
+
+        T, B, C = 3, 5, 7
+        x = torch.ones((T, B, C))
+        seq_lens = torch.tensor([3, 3, 2, 2, 1])
+        # set padding value so we can test equivalence
+        for b in range(B):
+            if seq_lens[b] < T:
+                x[seq_lens[b]:, b, :] = 0
+
+        eager_seq = pack_padded_pad_packed_script(x, seq_lens)
```
Comment (eellison) [inline test/test_jit.py]: reason not to do checkScript here instead ?

File: test/test_jit.py [hunk 2]
```diff
@@ -7736,6 +7736,27 @@ def forward(self, x, seq_lens):
         f = io.BytesIO()
         torch.onnx._export(m, (x, seq_lens), f, verbose=False)
 
+    def test_script_pack_padded_sequence(self):
+        from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
+
+        def pack_padded_pad_packed_script(x, seq_lens):
+            x = pack_padded_sequence(x, seq_lens)
+            x, _ = pad_packed_sequence(x)
```
Comment (eellison) [inline test/test_jit.py]: check both outputs here ?

**Review by eellison — APPROVED** [review]
LGTM. maybe add a couple more test inputs

File: torch/nn/utils/rnn.py [hunk 3]
```diff
@@ -254,7 +261,7 @@ def pack_padded_sequence(input, lengths, batch_first=False, enforce_sorted=True)
                       'sequence lengths. The tracer cannot track the data flow of Python '
                       'values, and it will treat them as constants, likely rendering '
                       'the trace incorrect for any other combination of lengths.',
-                      category=torch.jit.TracerWarning, stacklevel=2)
+                      stacklevel=2)
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Can't we make JIT ignore the category? It's not a good idea to remove it

File: torch/nn/utils/rnn.py [hunk 4]
```diff
@@ -310,12 +318,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0, total_le
...
-    padded_output, lengths = torch._C._VariableFunctions._pad_packed_sequence(
+    padded_output, lengths = _VF._pad_packed_sequence(
         sequence.data, sequence.batch_sizes, batch_first, padding_value, max_seq_length)
-    if sequence.unsorted_indices is not None:
+    unsorted_indices = sequence.unsorted_indices
+    if unsorted_indices is not None:
         batch_dim = 0 if batch_first else 1
-        return padded_output.index_select(batch_dim, sequence.unsorted_indices), \
-            lengths[sequence.unsorted_indices]
+        return padded_output.index_select(batch_dim, unsorted_indices), lengths[unsorted_indices]
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Why did you have to rewrite this code? Because of optional unwrapping?

File: torch/nn/utils/rnn.py [hunk 5]
```diff
@@ -310,12 +318,12 @@ def pad_packed_sequence(sequence, batch_first=False, padding_value=0.0, total_le
                              "total_length={} and max sequence length being {}"
                              .format(total_length, max_seq_length))
         max_seq_length = total_length
-    padded_output, lengths = torch._C._VariableFunctions._pad_packed_sequence(
+    padded_output, lengths = _VF._pad_packed_sequence(
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Is this because the JIT can't handle `torch._C._VariableFunctions`?

Comment (wanchaol) [reply torch/nn/utils/rnn.py]: Yes, we are using a wrapper for `_VariableFunctions` (which is _VF) for it to be accessible in JIT.

File: torch/nn/utils/rnn.py (hunk 4 above)
Comment (wanchaol) [reply torch/nn/utils/rnn.py]: Yes, our type refinement now is limited to TK_VAR only, we will need to expand it to a general case, see https://github.com/pytorch/pytorch/issues/23049

File: test/test_jit.py (hunk 1 above)
Comment (wanchaol) [reply test/test_jit.py]: `CheckScript` is also checking the string frontend and the string frontend could not handle the comments starting with `r`..

File: torch/nn/utils/rnn.py (hunk 3 above)
Comment (eellison) [reply torch/nn/utils/rnn.py]: category is hard to ignore atm because we don't have any way of typing `torch.jit.TracerWarning`

Comment (wanchaol) [reply torch/nn/utils/rnn.py]: yeah I tried that and it's kind of hard now, plus other callsites does not have this category either way.. not sure if someone deleted them before.

**Pull request closed without merging**